designcompose/src/main/java/com/android/designcompose/DesignView.kt

```diff
@@ -234,6 +235,65 @@ internal object DebugNodeManager {
     }
 }
 
+// Represents a key press event with optional meta keys. A DesignKeyEvent can be created with a
+// single character representing the key and a list of meta keys. It can also be created from a
+// list of javascript key codes, which is what Figma provides for an interaction with a key event
+// type trigger
+data class DesignKeyEvent(val key: Char, val metaKeys: List<DesignMetaKey>) {
+    companion object {
+        // Construct a DesignKeyEvent from a list of javascript key codes
+        fun fromJsKeyCodes(jsKeyCodes: List<Byte>): DesignKeyEvent {
+            var metaKeys: ArrayList<DesignMetaKey> = arrayListOf()
+            var key: Char = '0'
+            jsKeyCodes
+                .map { it.toInt() }
+                .forEach {
+                    when (it) {
+                        16 -> metaKeys.add(DesignMetaKey.MetaShift)
+                        17 -> metaKeys.add(DesignMetaKey.MetaCtrl)
+                        18 -> metaKeys.add(DesignMetaKey.MetaAlt)
+                        91 -> metaKeys.add(DesignMetaKey.MetaMeta)
+                        else -> key = it.toChar()
+                    }
+                }
+
+            return DesignKeyEvent(key, metaKeys)
+        }
+    }
+}
+
+// Manager to handle key event injects and listeners of key events
+private object KeyInjectManager {
+    private val keyListenerMap: HashMap<DesignKeyEvent, HashSet<(DesignKeyEvent) -> Unit>> =
+        HashMap()
+
+    // Inject a key event and notify listeners of the key event
+    fun injectKey(key: Char, metaKeys: List<DesignMetaKey>) {
+        val keyEvent = DesignKeyEvent(key, metaKeys)
+        val listeners = keyListenerMap[keyEvent]
+        listeners?.forEach { it(keyEvent) }
+    }
+
+    // Register a listener for a specific key event. This happens when a view with a key event
+    // trigger is composed.
+    fun addListener(keyEvent: DesignKeyEvent, setTriggeredKeyEvent: (DesignKeyEvent) -> Unit) {
+        if (keyListenerMap[keyEvent].isNullOrEmpty()) keyListenerMap[keyEvent] = HashSet()
+        keyListenerMap[keyEvent]?.add(setTriggeredKeyEvent)
+    }
+
+    // Remove a listener for a specific key event. This happens when a view with a key event trigger
+    // is removed from composition.
+    fun removeListener(keyEvent: DesignKeyEvent, setTriggeredKeyEvent: (DesignKeyEvent) -> Unit) {
+        val listeners = keyListenerMap[keyEvent]
+        listeners?.remove(setTriggeredKeyEvent)
+    }
+}
+
+// Public function to inject a key event
+fun DesignInjectKey(key: Char, metaKeys: List<DesignMetaKey>) {
```

Comment:
I'm also wondering if we want to expose this API. I think we're much better off if the connection between actions and the "keys" is explicit and can be in the JSON doc. If it's implicit like this, then we can't surface the events to the designer easily (because it could be any events coming in, not just the ones in the annotation).

Are you OK with not adding this to the API for now, and we'll revisit if we find out that we need it?

Reply:
I'm ok with it being internal, but this is called by the generated code, which so far calls public functions in DesignCompose.

Reply:
Oh, right, at some point we might want to name things like that differently (public symbol but should only be called by generated code, so we don't consider it a supported API), but since this is in-keeping with the current approach we're good.